### ISSUE TITLE
Support restricting a function in certain modules

### ISSUE THREAD

**Issue by zliu41** [ndmitchell/hlint#1392]
It's possible to restrict a function in all modules _except_ certain modules:

```
- functions:
  - {name: [fromIntegral], within: 'modules_where_fromIntegral_is_ok'}
```

But it doesn't seem possible to restrict a function only in certain modules. This would be a useful feature to add.

**Comment by ndmitchell** [comment]
So you declare a module where fromIntegral is banned, but by default it's allowed in most places? Seems a reasonable extension. Could be another field, or alternatively a prefix `!` in the within field to indicate it?

**Comment by zliu41** [comment]
> So you declare a module where fromIntegral is banned, but by default it's allowed in most places?

Exactly. `fromIntegral` is generally OK, but in certain modules it is unsafe and we want people to use `toIntegralSized`, which returns `Nothing` upon overflow/underflow.

> Could be another field, or alternatively a prefix ! in the within field to indicate it?

Yeah, something like that. Happy to open a PR

**Comment by ndmitchell** [comment]
Cool, PR welcome - ! prefix seems reasonable to me, but it's probably not the only thing that is reasonable.

**Comment by eyeinsky** [comment]
As part of ZuriHack I'm taking the :lock: on this issue :)

**Comment by eyeinsky** [comment]
In process of researching it but: `!` has special meaning in yaml for [tags](https://yaml.org/spec/1.2.2/#tags), so it might be that using it will require putting the module name(s) in quotes. Another way would be to add a separate key to put the exceptions in, e.g `disallow`.

We can bikeshed this while I look into where to add the part where the rule is enforced.

**Comment by zliu41** [comment]
Quoting it should work, I think - as in `within: "!Foo"`.

Or we can use a word that is the opposite of "within" - perhaps "outside"?

**Comment by michaelpj** [comment]
This would be nice, for large codebases you might have the capability to keep one part of your codebase clean of a particular function, and you want to list that subset of modules rather than listing out everything else.
